docs/content/sips/022-modularizing-spin-manifests.md

```diff
@@ -0,0 +1,264 @@
+
+[[trigger.http]]
+route = "/api/..."
+component = "sentiment-analysis"
+
+[[trigger.http]]
+route = "/..."
+component = "ui"
+
+[[trigger.http]]
+component = "kv-explorer"
+route = "/internal/kv-explorer/..."
+
+# The UI component is defined inline, because it's small.
+[component.ui]
+source = { url = "https://github.com/fermyon/spin-fileserver/releases/download/v0.0.1/spin_static_fs.wasm", digest = "sha256:650376c33a0756b1a52cad7ca670f1126391b79050df0321407da9c741d32375" }
+allowed_outbound_hosts = []
+files = [{ source = "../sentiment-analysis-assets", destination = "/" }]
+```
+
+```toml
+# [project]/sentiment-analysis/spin.toml
+
+# Sections and keys don't need to be scoped explicitly.
+[component]
+name = "sentiment-analysis"
+description = "Component-specific description"
+# Certain keys can be inherited from the application manifest
+authors.application = true
+version.application = true
+source = "target/spin-http-js.wasm"
+allowed_outbound_hosts = []
+exclude_files = ["**/node_modules"]
+key_value_stores = ["default"]
+ai_models = ["llama2-chat"]
+
+# Additional sections don't need to be namespaced under [component]
+[build]
+command = "npm run build"
+watch = ["src/**/*", "package.json", "package-lock.json"]
+```
+
+```toml
+# [project]/kv-explorer/Spin.toml
+
+[component]
+name = "kv-explorer"
+source = { url = "https://github.com/fermyon/spin-kv-explorer/releases/download/v0.10.0/spin-kv-explorer.wasm", digest = "sha256:65bc286f8315746d1beecd2430e178f539fa487ebf6520099daae09a35dbce1d" }
+allowed_outbound_hosts = ["redis://*:*", "mysql://*:*", "postgres://*:*"]
+# add or remove stores you want to explore here
+key_value_stores = ["default"]
+
+# Locally defined variables don't need to be namespaced manually.
+[variables]
+user = { required = true }
+password = { required = true }
+
+# Variables exposed to content (this needs bikeshedding!)
+[variables.exposed]
+kv_credentials = "{{ user }}:{{ password }}"
+```
+
+As can be seen in this example, this modular structure allows us to not only isolate aspects of individual components in their respective manifests, but also reduce repetition, since keys and sections don’t have to be explicitly scoped to the component anymore.
+
+## Details
+
+### Sub-manifests for components
+
+In this proposal, only components (and variables only accessible locally) can be defined in sub-manifests; triggers still need to be defined in the top-level manifest.
+
+There are multiple reasons for this constraint:
+
+1. Triggers are application-level concerns; they describe the application’s structure and high-level organization. As such, keeping them “at the surface” seems right.
+2. Trigger definitions tend to be very short, so moving them into sub-manifests individually as is done for components seems very boilerplate-y.
+3. Trigger definitions don’t come with their own assets, so any folder containing a sub-manifest for them would only ever contain the manifest, and nothing else.
```

Comment:
Mostly triggers don't come with component-related config, but we do have an oddity at the moment where WAGI execution has to be specified on the _trigger_ not the _component_ even though it's the component interface that mandates it.  (I _think_ this is to allow different trigger to set different argvs on the same component. Not sure though.)